src/main/java/org/jenkinsci/plugins/workflow/cps/EnvActionImpl.java

```diff
@@ -199,4 +203,29 @@ public Collection<EnvironmentContributor> getEnvironmentContributors() {
 
     }
 
+    @Extension public static class EnvActionImplPickleFactory extends SingleTypedPickleFactory<EnvActionImpl> {
+        @Override
+        protected Pickle pickle(EnvActionImpl object) {
+            return new EnvActionImplPickle();
+        }
+    }
+
+    /**
+     * Prevents multiple instances of {@link EnvActionImpl} from existing for a single Pipeline after a Jenkins restart
+     * in case {@code env} is serialized into the program.
+     */
+    private static class EnvActionImplPickle extends Pickle {
+        public ListenableFuture<?> rehydrate(FlowExecutionOwner owner) {
+            IOException cause = null;
+            try {
+                Queue.Executable executable = owner.getExecutable();
+                if (executable instanceof Run) {
+                    return Futures.immediateFuture(EnvActionImpl.forRun((Run)executable));
+                }
+            } catch (IOException e) {
+                cause = e;
+            }
+            return Futures.immediateFailedFuture(new RuntimeException("Unable to find Run for EnvActionImpl: " + owner, cause));
```

Comment:
I guess this doesn't need to be a `RuntimeException` any more (at first I was just `throw`ing the exception here). Maybe clearer to use `IOException`, idk.